Shaily_StataHw1_10:2.do

```diff
@@ -0,0 +1,234 @@
+	*extract just the grade
+	local varlist kiswahili english maarifa hisabati science uraia average
+	foreach subject in `varlist'{
+		replace `subject' = usubstr(`subject',-1,1)
+		}
+	
+	*unsure about this part
+	tempfile school`i'
+	save `school`i''
+	append using `school`i''
+	save psle_final, replace
+}
+
+/*******************************************************************************
+Q3: PAKISTAN NATIONAL ID
+*******************************************************************************/
+global excel_t21 "/Users/shailyacharya/Box/Econ490_Fall2022/Week4/04_assignment/data/Pakistan_district_table21.xlsx"
+*update the global
+
+clear
+*setting up an empty tempfile
+tempfile table21
+save `table21', replace emptyok
+
+*Run a loop through all the excel sheets (135) this will take 2-10 mins because it has to import all 135 sheets, one by one
+forvalues i=1/135 {
+	import excel "$excel_t21", sheet("Table `i'") firstrow clear allstring //import
+	display as error `i' //display the loop number
+
+	keep if regex(TABLE21PAKISTANICITIZEN1, "18 AND" )==1 //keep only those rows that have "18 AND"
+	*I'm using regex because the following code won't work if there are any trailing/leading blanks
+	*keep if TABLE21PAKISTANICITIZEN1== "18 AND" 
+	keep in 1 //there are 3 of them, but we want the first one
+	rename TABLE21PAKISTANICITIZEN1 table21
+
+	gen table=`i' //to keep track of the sheet we imported the data from
+	append using `table21' //adding the rows to the tempfile
+	save `table21', replace //saving the tempfile so that we don't lose any data
+}
+*load the tempfile
+use `table21', clear
+*fix column width issue so that it's easy to eyeball the data
+format %40s table21 B C D E F G H I J K L M N O P Q R S T U V W X Y Z AA AB AC
+
+local varlist table21 B C D E F G H I J K L M N O P Q R S T U V W X Y Z AA AB AC
+
+*left-aliging everything 
+foreach v of varlist `varlist' {
+    
+        local f : format `v'
+        
+        local ff = regexr("`f'", "%~?\-?","")
+        
+        if regexm("`f'", "%.+s") local cmd cap format %-`ff' `v'
+        
+        if "`: value label `v''" != "" & "`label'" != "" ///
+            local cmd cap format %-`ff' `v'
+        
+        if "`all'" != "" local cmd cap format %-`ff' `v'
+        
+        `cmd'
+        
+        local ff : format `v'
+        if "`ff'" != "`f'" local vlist `vlist' `v'
+
+    }
+    
+    if "`vlist'" != "" des `vlist'
+	
+sort table 
+order table, before(table21)
+drop table21 //this column is useless
+drop AC //this column has no observations
+
+*now combine columns and rename according to how they appear in PDF
+*could I make any of the below into a loop?
+gen totalpop = (B + C)
+gen cni_obtained = (D+E)
+gen no_cni = (F+G)
+
+gen m_totalpop = (H+I)
+gen m_cni_obtained = (J+K)
+gen m_no_cni = (L+M)
+
+gen f_totalpop = (N+O)
+gen f_cni_obtained = (P+Q)
+gen f_no_cni = (R+S)
+
+gen t_totalpop = (T+U)
+gen t_cni_obtained = (V+W)
+gen t_no_cni = (AA+AB+Z+X+Y)
+
```

Comment:
Good job on setting up question 3!! However, there were a few values (see cni_obtained 26, 32, and 33) that seem to be on the higher side and may have been miscalculated. 

I think it might be easier to bring all the data points into a single column and then reshape it back to the wide format to eliminate missing values. Because then we do not have to manually calculate values.